CorsixTH/Lua/game_ui.lua

```diff
@@ -750,6 +752,30 @@ function GameUI:onMouseWheel(x, y)
   return UI.onMouseWheel(self, x, y)
 end
 
+--! Announcements to the player.
+--!param msgs (array of string) Messages to select from.
+--!param rnd_frac (optional float in range (0, 1]) Fraction of times that the
+--    call actually says something.
+--!param priority (optional string - Low, Normal, High, Critical) Priority of announcement
+--!return (boolean) Whether a message was given to the user.
+function GameUI:makeAnnouncement(msgs, rnd_frac, priority)
+  if not priority or not AnnouncementPriority[priority] then priority = "Normal" end
+  assert(priority, "Invalid announcement priority")
```

Comment:
One line above you say `if  not priority .... then priority = "Normal" end`
How can this assert ever trigger?

Reply:
I'm not sure how it could trigger, but I think that about other asserts too.

Reply:
> I'm not sure how it could trigger, but I think that about other asserts too.

All asserts are for protecting against things that are not supposed to be happening. The point is indeed what you say, we mostly don't know if they can happen. In such a case the `assert` is kept, mostly as a safety measure. (Basically, it's too much work or too complicated to proof the assert never triggers, keeping it is cheaper.)

However, if you can easily proof that you have code equivalent to `assert(true)`, you should remove it, as it protects against an impossibility, and only clutters the code.

There are 3 cases here:
- `priority` evaluates to `false` or `nil`: In this case, the `if` jumps into the `then`, and sets the variable to a string which does not evaluate to `false` or `nil`. This case thus doesn't trigger the assert.
- `priority` evaluates to `true`, and `AnnouncementPriority[priority]` evaluates to `false`/`nil`. In this case too, the `if` fixes the value like the previous case, assert again doesn't trigger.
-  `priority` evaluates to `true`, and `AnnouncementPriority[priority]` evaluates to `true`. The `if` doesn't do anything. Since `priority` stays the same, the assert doesn't trigger.

So for all possible cases that I see, the assert doesn't trigger. My conclusion therefore is that the assert isn't useful.

Note that this can be decided only because you can here locally reason for assert not triggering. This is quite cheap to do. If code becomes more complicated, you can reach the point where it is not clear if a problem can happen. In such a case, you would add the assert again at that time.